contrib/golang/filters/http/source/go/pkg/http/type.go

```diff
@@ -82,6 +82,12 @@ func (h *httpHeaderMap) Values(key string) []string {
 }
 
 func (h *httpHeaderMap) Set(key, value string) {
+	// Get all header values first before setting a value, since the set operation may not take affects immediately
+	// when it's invoked in a Go thread, instead, it will post a callback to run in the envoy worker thread.
+	// Otherwise, we may get outdated values.
```

Comment:
Sorry, but I don't get the logic. Will `cAPI.HttpCopyHeaders` waiting all pending modificattions to be completed and return latest values?

Reply:
Nope, `cAPI.HttpCopyHeaders` only delivers the pending modification, won't wait for it to be completed.

So, we might get outdated values in a following `Get` call after `Set`, i.e.
```
header.Set("foo", "bar") // the pending modification may haven't take affects yet.
header.Get("foo") // might return the outdated value of foo
```

Also, updated the comment clearly.

Reply:
> Nope, cAPI.HttpCopyHeaders only delivers the pending modification, won't wait for it to be completed.

Then how can the `cAPI.HttpCopyHeaders` prevent you from getting outdated value?

Reply:
We copy the headers from Envoy first, then, the values on the Go side are in sync.
After that, we set the values on both the Go side and the Envoy side, keeping the values always synced.

Actually, it's not real-time synced, since changes might delay on the Envoy side, but we assume the changes must happen, which means we guarantee eventual consistency.